File: BitonicSort/bitonic_sort.c
#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>
#include <omp.h>
#include <math.h>
#include <time.h>
#define COMPARE(A, B) ((A > B)? 1: -1)
#define UP -1
#define DOWN 1
#define PARALLEL_THRESHOLD 16384 

void copy_array(int* source, int* destination, int size);
bool is_power_of_two(int n);
void print_arr(int* arr, int size);
void bitonic_sort(int* arr, int start, int cnt, int dir);
void bitonic_merge(int* arr, int start, int cnt, int dir);
void bitonic_sort_p(int* arr, int start, int cnt, int dir);
void bitonic_merge_p(int* arr, int start, int cnt, int dir);
void swap(int*arr, int a, int b);

int main(int argc, char* argv[]) {

	clock_t start, end, start_p, end_p;
	double execution_time, execution_time_p;

	if (argc < 2) {
		fprintf(stderr, "Please insert array size.\n");
		return 1;
	}

	/* Get array size from command line */
	int array_size = strtol(argv[1], NULL, 10);
	if (!is_power_of_two(array_size)) {
		fprintf(stderr, "Please insert array size with the power of two.\n");
		return 1;
	}

	/* Generate array with random numebrs */
	int* arr = malloc(array_size*sizeof(int));
	srand(time(NULL));
	for (int i = 0; i < array_size; i++) {
		arr[i] = rand() % 30;
	}
	int* arr_p = malloc(array_size*sizeof(int));
	copy_array(arr, arr_p, array_size);

	/* printf("Original Array: ");
	print_arr(arr, array_size); */

	/* Apply bitonic sort */
	start = clock();
	bitonic_sort(arr, 0, array_size, UP);
	end = clock();

        /* printf("Sorted Array: ");
        print_arr(arr, array_size); */
	
	execution_time = ((double) end - start) / CLOCKS_PER_SEC;
	printf("Execution time: %f seconds\n", execution_time);

        /* Apply parallelized bitonic sort */
        start_p = clock();
     	#pragma omp parallel
	{
		#pragma omp single
		{
		bitonic_sort_p(arr_p, 0, array_size, UP);
		}
	}
        end_p = clock();

        /* printf("Sorted Array using parallelism: ");
        print_arr(arr_p, array_size); */

        execution_time_p = ((double) end_p - start_p) / CLOCKS_PER_SEC;
        printf("Execution time using parallelism: %f seconds\n", execution_time_p);
	printf("Speedup: %f\n", execution_time / execution_time_p);
	
	free(arr);
	free(arr_p);
}

void copy_array(int* source, int* destination, int size) {
    for (int i = 0; i < size; i++) {
        destination[i] = source[i];
    }
}

bool is_power_of_two(int n) {
	return n > 0 && (n & (n - 1)) == 0;
}


void print_arr(int* arr, int size) {
	for (int i = 0; i < size; i++) {
		printf("%d ", arr[i]);
	}
	printf("\n");
}

void bitonic_sort(int* arr, int start, int cnt, int dir) {
	if (cnt <= 1) return;
	int size = cnt / 2;
	
	/* Sort the first half in increasing order */
	bitonic_sort(arr, start, size, UP);
	/* Sort the second half in decreasing order */
	bitonic_sort(arr, start + size, size, DOWN);

	/* Merge the two sorted parts */
	bitonic_merge(arr, start, cnt, dir);
}

void bitonic_merge(int* arr, int start, int cnt, int dir) {
	if (cnt <= 1) return;
	int size = cnt / 2;

	/* Compare and swap first and second half of the array */
	for (int i = start; i < start + size; i++) {
		if (dir != COMPARE(arr[i], arr[i + size])) {
			swap(arr, i, i + size);
		}
	}
	
	bitonic_merge(arr, start, size, dir);
	bitonic_merge(arr, start + size, size, dir);
}

void bitonic_sort_p(int* arr, int start, int cnt, int dir) {
	if (cnt <= PARALLEL_THRESHOLD) {
		bitonic_sort(arr, start, cnt, dir);
		return;
	}

	int size = cnt / 2;

	#pragma omp task
	bitonic_sort_p(arr, start, size, UP);
	
	#pragma omp task
	bitonic_sort_p(arr, start + size, size, DOWN);

	#pragma omp taskwait
	bitonic_merge_p(arr, start, cnt, dir);
}

void bitonic_merge_p(int* arr, int start, int cnt, int dir) {
    	if (cnt <= PARALLEL_THRESHOLD) {
	    	bitonic_merge(arr, start, cnt, dir);
		return;
	}   
        
	int size = cnt / 2;

	/* Compare and swap first and second half of the array */
	#pragma omp parallel for
        for (int i = start; i < start + size; i++) {
                if (dir != COMPARE(arr[i], arr[i + size])) {
                        swap(arr, i, i + size);
                }
        }

	#pragma omp task
	bitonic_merge(arr, start, size, dir);
	#pragma omp task
	bitonic_merge(arr, start + size, size, dir);
	#pragma omp taskwait
}

void swap(int*arr, int a, int b) {
	int temp = arr[b];
	arr[b] = arr[a];
	arr[a] = temp;
}

File: BitonicSort/bitonic_sort.rs
use rayon::prelude::*;
use std::env;
use std::time::Instant;

const UP: i32 = -1;
const DOWN: i32 = 1;
const PARALLEL_THRESHOLD: usize = 16384;

fn main() {
    // Get array size from command line
    let args: Vec<String> = env::args().collect();
    if args.len() < 2 {
        eprintln!("Please insert array size.");
        return;
    }

    let array_size: usize = args[1].parse().expect("Invalid number");

    if !is_power_of_two(array_size) {
        eprintln!("Please insert array size that is a power of two.");
        return;
    }

    // Generate array with random numbers
    let mut arr: Vec<i32> = (0..array_size).map(|_| rand::random::<i32>() % 30).collect();
    //print_array(&arr);
    let mut arr_p = arr.clone();

    // Apply bitonic sort
    let start = Instant::now();
    bitonic_sort(&mut arr, array_size, UP);
    let duration = start.elapsed();
    //print_array(&arr);
    println!("Execution time: {:.6} seconds", duration.as_secs_f64());

    // Apply parallelized bitonic sort
    let start_p = Instant::now();
    bitonic_sort_p(&mut arr_p, array_size, UP);
    let duration_p = start_p.elapsed();
    //print_array(&arr_p);
    println!("Execution time using parallelism: {:.6} seconds", duration_p.as_secs_f64());
    println!("Speedup: {:.6}", duration.as_secs_f64() / duration_p.as_secs_f64());
}

// Function to print the array
fn print_array(arr: &[i32]) {
    for &item in arr.iter() {
        print!("{} ", item);
    }
    println!();
}

// Check if a number is a power of two
fn is_power_of_two(n: usize) -> bool {
    n > 0 && (n & (n - 1)) == 0
}

// Bitonic sort
fn bitonic_sort(arr: &mut [i32],cnt: usize, dir: i32) {
    if cnt <= 1 {
        return;
    }

    let size = cnt / 2;
    let (first, second) = arr.split_at_mut(size);
    bitonic_sort(first, size, UP);
    bitonic_sort(second, size, DOWN);
    bitonic_merge(arr, cnt, dir);
}

// Bitonic merge
fn bitonic_merge(arr: &mut [i32], cnt: usize, dir: i32) {
    if cnt <= 1 {
        return;
    }

    let size = cnt / 2;
    for i in 0..size {
        if (arr[i] > arr[i + size]) == (dir == UP) {
            arr.swap(i, i + size);
        }
    }

    let (first, second) = arr.split_at_mut(size);
    bitonic_merge(first, size, dir);
    bitonic_merge(second, size, dir);
}

// Parallel bitonic sort
fn bitonic_sort_p(arr: &mut [i32], cnt: usize, dir: i32) {
    if cnt <= PARALLEL_THRESHOLD {
        bitonic_sort(arr, cnt, dir);
        return;
    }

    let size = cnt / 2;
    let (first, second) = arr.split_at_mut(size);

    rayon::join(
        || bitonic_sort_p(first, size, UP),
        || bitonic_sort_p(second, size, DOWN),
    );

    bitonic_merge_p(arr, cnt, dir);
}

// Parallel bitonic merge
fn bitonic_merge_p(arr: &mut [i32], cnt: usize, dir: i32) {
    if cnt <= PARALLEL_THRESHOLD {
        bitonic_merge(arr, cnt, dir);
        return;
    }

    let size = cnt / 2;

    let swap_indices: Vec<(usize, usize)> = arr.par_iter()
        .enumerate()
        .take(size)
        .filter_map(|(i, _)| {
            let idx1 = i;
            let idx2 = i + size;
            if (arr[idx1] > arr[idx2]) == (dir == UP) {
                Some((idx1, idx2))
            } else {
                None
            }
        }).collect();

    // Perform the swaps outside the initial borrow
    for (idx1, idx2) in swap_indices {
        arr.swap(idx1, idx2);
    }

    let (first, second) = arr.split_at_mut(size);
    rayon::join(
        || bitonic_merge_p(first, size, dir),
        || bitonic_merge_p(second, size, dir),
    );
}

